homeassistant/components/sensor/history_average.py

```diff
@@ -0,0 +1,342 @@
+                              name, unit)])
+    return True
+
+
+class HistoryAverageSensor(Entity):
+    """HistoryAverage sensor which calculates time weighted averages."""
+
+    @asyncio.coroutine
+    def async_get_history(self):
+        """Pull states from recorder history."""
+        yield from self.async_update_period()
+        start, _end = self._period
+
+        self._history = []
+
+        # get state changes history between start and now
+        states = history.state_changes_during_period(
+            self._hass, start, None, self._entity_id)
+        states = states.get(self._entity_id)
+        if states:
+            self._history.extend(states)
+
+    @asyncio.coroutine
+    def asysnc_trim_history(self):
+        """Remove items which are not relevent to the current period."""
+        yield from self.async_update_period()
+        start, _end = self._period
+        start_timestamp = dt_util.as_timestamp(start)
+
+        def after_start(state, timestamp):
+            """Test if the State last changed after the timestamp."""
+            return state.last_changed.timestamp() >= timestamp
+
+        # remove items before the current period's starting point,
+        # keeping the first item before the starting period to enable
+        # calculation of the state between start -> next state change
+        for index, state in enumerate(self._history[:]):
+            if after_start(state, start_timestamp):
+                break
+            else:
+                # remove if the next item is also before start
+                next_index = index + 1
+                if (next_index < len(self._history) and
+                        (not after_start(self._history[next_index],
+                                         start_timestamp))):
+                    del self._history[index]
+
+    def get_period(self):
+        """Return current period, used for testing."""
+        return self._period
+
+    def __init__(
+            self, hass, entity_id, start, end, duration,
+            name, unit):
+        """Initialize the HistoryAverage sensor."""
+        self._hass = hass
+        self._entity_id = entity_id
+        self._duration = duration
+        self._start = start
+        self._end = end
+        self._name = name
+        self._unit_of_measurement = unit
+
+        # set defaults at init
+        now = dt_util.utcnow()
+        self._period = (now, now)
+        self._history = []
+        self._state = 0
+
+    @asyncio.coroutine
+    def async_added_to_hass(self):
+        """Initialize data and register callbacks."""
+        # load data from recorder and do initial calculation
+        yield from self.async_get_history()
+        yield from self.async_update()
+
+        @callback
+        def state_listener(entity, old_state, new_state):
+            """Handle the sensor state changes."""
+            self._history.append(new_state)
+            sorted(self._history, key=attrgetter('last_changed'))

```

Comment:
Sorted does not sort in-place, it returns a new list. Also, when will the order never be old…new? I think that you can remove this.

Reply:
My concern is that this can be called asynchronously, correct? So, at least in theory, there's a chance that the state_listener will get invoked in the wrong order. Probably unlikely, unless there's a rapid fire number of events.

Thanks for the catch about not actually saving the result of the sort.